### PR TITLE
[ENH] Add support for extension arrays to expand_grid

### LINKED ISSUE #1121: `expand_grid` for extension types
- Add a dispatch for pandas extension arrays

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

- add support for extension arrays to expand_grid

**This PR resolves #1121**

# PR Checklist

Please ensure that you have done the following:

1. [ ] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.md`.

3. [x] Add a line to `CHANGELOG.md` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

# Automatic checks

There will be automatic checks run on the PR. These include:

- Building a preview of the docs on Netlify
- Automatically linting the code
- Making sure the code is documented
- Making sure that all tests are passed
- Making sure that code coverage doesn't go down.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by ericmjl** [comment]
🚀 Deployed on https://deploy-preview-1122--pyjanitor.netlify.app

File: examples/notebooks/expand_grid.ipynb [hunk 1]
```diff
@@ -1125,9 +1123,14 @@
...
-   "version": "3.9.9"
+   "version": "3.9.10"
   },
-  "orig_nbformat": 2
+  "orig_nbformat": 2,
+  "vscode": {
+   "interpreter": {
+    "hash": "d4d1e4263499bec80672ea0156c357c1ee493ec2b1c70f0acce89fc37c4a6abe"
+   }
+  }
```
Comment (Zeroto521) [inline examples/notebooks/expand_grid.ipynb]: I suggest deleting the personal information `vscode` section.
If I run this .ipynb in my environment, this information will change again.

File: janitor/utils.py [hunk 2]
```diff
@@ -150,13 +149,12 @@ def _sub_expand_grid(value, grid_index, key):  # noqa: F811
 def _sub_expand_grid(value, grid_index, key):  # noqa: F811
     """
     Expands the Index based on `grid_index`.
-
     Returns a dictionary.
     """
     name = value.name
     if not name:
         name = 0
```
Comment (Zeroto521) [inline janitor/utils.py]:
```suggestion
    name = value.name or 0
```

Comment (samukweku) [reply janitor/utils.py]: Do you mind explaining the suggestion? The if stmt seems clear, I would however love to understand the suggestion. If value.name is None the or part kicks in? Is that part of python's grammar where the if clause is implied?

Comment (Zeroto521) [reply janitor/utils.py]: It’s a trick like “x if conditions else y” but more convenient.
When “conditions” is “x” (x if x else y) equal to “x or y”.

**Comment by Zeroto521** [comment]
why was examples/notebooks/expand_grid.ipynb deleted?

**Comment by samukweku** [comment]
> why was examples/notebooks/expand_grid.ipynb deleted?

did I misunderstand your comment regarding the notebook? I intend pulling it back in another PR. Also we moved the notebooks to a different repo, so the deletion does not incur any loss.

**Comment by Zeroto521** [comment]
> did I misunderstand your comment regarding the notebook?

I have suggested removing those configs to avoid frequently changing like ‘vscode’ which I pointed out.
It was generate by vscode and uncorrelated to jupyter and example.

> I intend pulling it back in another PR. Also we moved the notebooks to a different repo, so the deletion does not incur any loss.

I just thought delete this file would change a lot for this PR.
There would have another PR. We could do this action in that new PR to keep a small step forward.

**Review by Zeroto521 — APPROVED** [review]
Thanks for doing that.
After updating the changelog file we can merge this PR.
It seems the changelog commit was missed due to 'git reset'.

**Pull request merged**